zephyr/src/supp_api.c

```diff
@@ -65,27 +69,155 @@ static int send_wpa_supplicant_dummy_event(void)
 	return send_wpa_supplicant_event(&msg);
 }
 
+/* Lifted from zephyr shell_utils.c to handle escapes */
+static inline uint16_t supp_strlen(const char *str)
+{
+	return str == NULL ? 0U : (uint16_t)strlen(str);
+}
 
-static inline struct wpa_supplicant * get_wpa_s_handle(const struct device *dev)
+static char make_argv(char **ppcmd, uint8_t c)
 {
-	struct wpa_supplicant *wpa_s = NULL;
-	int ret = k_sem_take(&wpa_supplicant_ready_sem, K_SECONDS(2));
+	char *cmd = *ppcmd;
+	char quote = 0;
 
-	if (ret) {
-		wpa_printf(MSG_ERROR, "%s: WPA supplicant not ready: %d", __func__, ret);
-		return NULL;
-	}
+	while (1) {
+		c = *cmd;
 
-	k_sem_give(&wpa_supplicant_ready_sem);
+		if (c == '\0') {
+			break;
+		}
 
-	wpa_s = wpa_supplicant_get_iface(global, dev->name);
-	if (!wpa_s) {
-		wpa_printf(MSG_ERROR,
-		"%s: Unable to get wpa_s handle for %s\n", __func__, dev->name);
-		return NULL;
+		if (quote == c) {
+			memmove(cmd, cmd + 1, supp_strlen(cmd));
+			quote = 0;
+			continue;
+		}
+
+		if (quote && c == '\\') {
+			char t = *(cmd + 1);
+
+			if (t == quote) {
+				memmove(cmd, cmd + 1,
+						supp_strlen(cmd));
+				cmd += 1;
+				continue;
+			}
+
+			if (t == '0') {
+				uint8_t i;
+				uint8_t v = 0U;
+
+				for (i = 2U; i < (2 + 3); i++) {
+					t = *(cmd + i);
+
+					if (t >= '0' && t <= '7') {
+						v = (v << 3) | (t - '0');
+					} else {
+						break;
+					}
+				}
+
+				if (i > 2) {
+					memmove(cmd, cmd + (i - 1),
+						supp_strlen(cmd) - (i - 2));
+					*cmd++ = v;
+					continue;
+				}
+			}
+
+			if (t == 'x') {
+				uint8_t i;
+				uint8_t v = 0U;
+
+				for (i = 2U; i < (2 + 2); i++) {
+					t = *(cmd + i);
+
+					if (t >= '0' && t <= '9') {
+						v = (v << 4) | (t - '0');
+					} else if ((t >= 'a') &&
+						   (t <= 'f')) {
+						v = (v << 4) | (t - 'a' + 10);
+					} else if ((t >= 'A') && (t <= 'F')) {
+						v = (v << 4) | (t - 'A' + 10);
+					} else {
+						break;
+					}
+				}
+
+				if (i > 2) {
+					memmove(cmd, cmd + (i - 1),
+						supp_strlen(cmd) - (i - 2));
+					*cmd++ = v;
+					continue;
+				}
+			}
+		}
+
+		if (!quote && isspace((int) c)) {
+			break;
+		}
+
+		cmd += 1;
 	}
+	*ppcmd = cmd;
 
-	return wpa_s;
+	return quote;
+}
+
+
+char supp_make_argv(size_t *argc, const char **argv, char *cmd,
+		       uint8_t max_argc)
+{
+	char quote = 0;
+	char c;
+
+	*argc = 0;
+	do {
+		c = *cmd;
+		if (c == '\0') {
+			break;
+		}
+
+		if (isspace((int) c)) {
+			*cmd++ = '\0';
+			continue;
+		}
+
+		argv[(*argc)++] = cmd;
+		if (*argc == max_argc) {
+			break;
+		}
+		quote = make_argv(&cmd, c);
+	} while (true);
+
+	return quote;
+}
+
+
+int call_wpa_cli (const char *fmt, ...)
```

Comment:
Should this be part of the `wpa_cli_zephyr.h` API as, for example, `z_wpa_cli_cmd_v`?

Reply:
I would appreciate something like this to be exposed for ease of use. :)